hnn_core/extracellular.py

```diff
@@ -0,0 +1,551 @@
+                                 f'{len(positions)} for trial {ii}')
+            for jj in range(len(voltages[ii])):
+                if len(times) != len(voltages[ii][jj]):
+                    raise ValueError('length of times and voltages must match,'
+                                     f' got {len(times)} and '
+                                     f'{len(voltages[ii][jj])} for trial{ii}, '
+                                     f'channel {jj}')
+
+        self.positions = positions
+        self.sigma = sigma
+        self.method = method
+        self.min_distance = min_distance
+
+        self._times = times
+        self._data = voltages
+
+    def __getitem__(self, trial_no):
+        try:
+            if isinstance(trial_no, int):
+                return_data = [self._data[trial_no]]
+            elif isinstance(trial_no, slice):
+                return_data = self._data[trial_no]
+            elif isinstance(trial_no, (list, tuple)):
+                return_data = [self._data[trial] for trial in trial_no]
+            else:
+                raise TypeError(f'trial index must be int, slice or list-like,'
+                                f' got: {trial_no} which is {type(trial_no)}')
+        except IndexError:
+            raise IndexError(f'the data contain {len(self)} trials, the '
+                             f'indices provided are out of range: {trial_no}')
+        return ExtracellularArray(self.positions, sigma=self.sigma,
+                                  method=self.method, times=self.times,
+                                  voltages=return_data)
+
+    def __repr__(self):
+        class_name = self.__class__.__name__
+        msg = (f'{len(self.positions)} electrodes, sigma={self.sigma}, '
+               f'method={self.method}')
+        if len(self._data) > 0:
+            msg += f' | {len(self._data)} trials, {len(self.times)} times'
+        else:
+            msg += ' (no data recorded yet)'
+        return f'<{class_name} | {msg}>'
+
+    def __len__(self):
+        return len(self._data)
+
+    @property
+    def times(self):
+        return self._times
+
+    @property
+    def voltages(self):
+        return self._data
+
+    @property
+    def sfreq(self):
+        """Return the sampling rate of the extracellular data."""
+        if len(self.times) == 0:
+            return None
+        elif len(self.times) == 1:
+            raise RuntimeError('Sampling rate is not defined for one sample')
+
+        dT = np.diff(self.times)
+        Tsamp = np.median(dT)
+        if np.abs(dT.max() - Tsamp) > 1e-3 or np.abs(dT.min() - Tsamp) > 1e-3:
+            raise RuntimeError(
+                'Extracellular sampling times vary by more than 1 us. Check '
+                'times-attribute for errors.')
+
+        return 1000. / Tsamp  # times are in in ms
+
+    def _reset(self):
+        self._data = list()
+        self._times = list()
+
+    def get_data(self, return_times=False):
```

Comment:
I am very confused with all the different ways the data/voltages can be accessed. There seems to be slicing/private attributes/property/get_data. Zen of Python says:

```sh
There should be one-- and preferably only one --obvious way to do it.
```

Feel some of it will automatically simplify if we combine into a single class

Reply:
Agreed, but see above. I don't think a single class will help with this... An executive decision on internal data containers might, though.

Reply:
If `voltages` and `times` are numpy arrays, the `get_data` method loses its meaning.